### ISSUE TITLE
Prometheus is not able to scrape any of the running pods

### ISSUE THREAD

**Issue by Hyaxia** [hashicorp/consul-k8s#1205]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request. Searching for pre-existing feature requests helps us consolidate datapoints for identical requirements into a single place, thank you!
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

---

### Overview of the Issue

I'm pretty new to consul so I might be getting things wrong, but I've wanted to test it out on my laptop (MacBook m1 if its relevant since there are lots of problems with m1s these days).
I started a minikube cluster, then, using terraform and helm I've created few deployments that include consul and other two web apps.
When I tried to see what metrics prometheus got I saw that he failed to scrape any of the pods running on the cluster.
Each time he fails with an error: `"INVALID" is not a valid start token`.

### Reproduction Steps

The entire project is on my [GitHub](https://github.com/Hyaxia/terraform-minikube/tree/feature/helmify) (notice that you have to use the branch `feature/helmify`).
The README of the project contains all of the steps to recreate the environment that I'm working with.

After running `terraform apply` if we run `kubectl --namespace consul port-forward service/consul-consul-ui 18500:80 --address 0.0.0.0` to check the consul dashboard and we go into the topology of the backend service we will see that the metrics are loading like [that](https://i.stack.imgur.com/QWezP.png)

If we run `kubectl --namespace consul port-forward service/prometheus-server 9090:80 --address 0.0.0.0` and go to the prometheus api on `http://localhost:9090/api/v1/targets` we will see all of the targets that he scrapes and the health of the backend pods will be "down". 
If we try to create a service that exposes port 20200 on one of the pods and then port forward that to perform the request to `/metrics` ourselves we will see that the metrics do exist on that route.

### Logs

I could not find any helpful logs on the prometheus server or the pods.

### Expected behavior

Prometheus should be able to scrape the pods.

### Environment details

- Kubernetes version: v1.23.3
- Minikube version: v1.25.2
- Macbook M1

**Comment by Hyaxia** [comment]
Is there anything missing? something I can add?

**Comment by jmurret** [comment]
Hi @Hyaxia,  have you been able to successfully go through the [docs on Prometheus](https://www.consul.io/docs/k8s/connect/observability/metrics) as well as the [Learn Guide on Prometheus](https://learn.hashicorp.com/tutorials/consul/service-mesh-observability?in=consul/kubernetes-features)?  Did you have any issues completing the Learn Guide successfully?
We also have a couple of [automated acceptance tests](https://github.com/hashicorp/consul-k8s/blob/a0f06f9e32e8e202c73b039100aab0283090fb77/acceptance/tests/metrics/metrics_test.go#L91) that ensure the metrics feature works across a matrix of kubernetes distributions.   They can serve as a good reference.

I think starting here and seeing if you run into obstacles with one of our tutorials would be the best starting point.  From there you will be better equipped to inspect the app that you have created to roll this out.   Please let me know if you questions and let us know how the tutorials go.

**Comment by Hyaxia** [comment]
Hi @jmurret, yes I did go through both of the guides and configured the project accordingly.
I tried to access the `/metrics` route to see if I get any metrics by myself and I was able to do so.
If you could try to run the project that is on GitHub and see for yourself if its works that would be great.
If it does work, then the problem is with my laptop, if it doesn't then the problem is with my configuration.

**Comment by david-yu** [comment]
@Hyaxia Are you using the Consul on Kubernetes Helm chart? If so could you provide your values.yaml overrides file? The metrics scraping feature through a merged metrics endpoint is something that is custom built int our project and not something available with Consul alone.

**Comment by Hyaxia** [comment]
@david-yu yes I'm using the helm chart.
The values are:
```yaml
resource "helm_release" "consul" {
  name       = "consul"
  repository = "https://helm.releases.hashicorp.com"
  chart      = "consul"
  version    = "0.39.0"
  namespace  = "consul"

  # global
  set {
    name  = "global.metrics.enabled"
    value = true
  }

  # server
  set {
    name  = "server.replicas"
    value = 1
  }

  # controller
  set {
    name  = "controller.enabled"
    value = true
  }
  
  # connectInject
  set {
    name  = "connectInject.enabled"
    value = true
  }
  set {
    name  = "connectInject.default"
    value = false
  }
  set {
    name  = "connectInject.metrics.defaultEnabled"
    value = true
  }
  set {
    name  = "connectInject.metrics.defaultEnableMerging"
    value = true
  }

  # prometheus
  set { # this configuration is for deploying prometheus for non production use cases
    name  = "prometheus.enabled"
    value = true
  }

  # ui
  set {
    name  = "ui.enabled"
    value = true
  }
  set {
    name  = "ui.metrics.enabled"
    value = true
  }
  set {
    name  = "ui.metrics.provider"
    value = "prometheus"
  }
  set {
    name  = "ui.metrics.baseUrl"
    value = "http://prometheus-server"
  }
}

```

You can view the entire configuration in the [GitHub](https://github.com/Hyaxia/terraform-minikube/blob/feature/helmify/main.tf) project

**Comment by Hyaxia** [comment]
The problem was that I while testing stuff I defined a route in my web server that takes all of the requests for paths that start with the root and have no other back slashes, this caused the requests from prometheus scraper to be handled by this custom route instead of the agent installed on the container. Fixed. Sorry for the time wasted on my mistakes :upside_down_face:

**Comment by ADustyOldMuffin** [comment]
@jmurret So I'm having the same issue, we're scraping all pods via an outside instance of Prometheus, but for Consul to be scraped it must have `?format=prometheus` provided which you cannot do in K8s to my knowledge.

We're using 1.6.X version of Consul which doesn't allow for any of the UI options (I've had to modify the helm chart to get it to work which I detailed in issue #1421). I'm not sure if older versions of Consul are going to be supported, but if anyone shows up from Google that's why and there's the issue to track.
